### ISSUE TITLE
Connection leak problem

### ISSUE THREAD

**Issue by ngmenyen** [DICE-UNC/jargon#231]
Hi, all

I'm getting errors messages from a machine that has installed metalnx and jargon in it. It is displaying a message that a potential connection leak was found like this one below:

> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:95 - ********  connection is:irods://rods.tempZone@tpa-eld6555:1247/http-bio-8080-exec-45/1487182980282, will attempt to disconnect and shut down any restart thread
> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:98 - **************************************************************************************
> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:91 - **************************************************************************************
> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:92 - ********  WARNING: POTENTIAL CONNECTION LEAK  ******************
> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:93 - ********  finalizer has run and found a connection left opened, please check your code to ensure that all connections are closed
> 2017-02-15 17:01:12 ERROR IRODSMidLevelProtocol:94 - ********  IRODSCommands is:AbstractIRODSMidLevelProtocol [irodsConnection=irods://test2.tempZone@tpa-eld6555:1247/http-bio-8080-exec-45/1487182979688, irodsProtocolManager=org.irods.jargon.core.connection.IRODSSimpleProtocolManager@71aa1758, irodsServerProperties=IRODS server properties

I was wondering what is the scenario for this to happen. I was doing upload with a lot of files when I got this message. Can anyone help me with an answer?

Best regards,
Ng Men Yen

**Comment by michael-conway** [comment]
Are you calling REST, or calling Jargon API directly?

That message means that the finalizer ran and found that a connection was still open, so it closes it behind the scenes for you, but is warning that connections are not being closed in a process.  

You can use methods in IRODSAccessObjectFactory or IRODSFileSystem to close, either giving an IRDOSAccount, or with no account parameter, that will close all connections in a thread.

**Comment by ngmenyen** [comment]
I am using Jargon API. We normally use the IRODSAccessObjectFactory  or IRODSFileSystem when we logout of the session, should we use it after every operation in jargon? Doesn't the close method end the session between the logged user and iRODS?

**Comment by michael-conway** [comment]
For web applications, it's recommended that the connection not be held in
the session.  I think that's probably OK, but will hold an agent open.  Are
you seeing this as a result of session time-outs?

For REST and WebDav, I'll be adding a connection cache behind Jargon so
that conns are pooled between invocations, but I typically use a filter
pattern to close the connection after an HTTP request, just to not set up
idle agents on iRODS.

We can talk more directly if you want to compare notes!
MC

**Comment by dkocher** [comment]
This is the same issue as described in #199 with a proposed fix in #213.

**Comment by michael-conway** [comment]
HTTP requests should follow a session per request model

**Comment by michael-conway** [comment]
Can you PM me via my UNC email?  mconway@unc.edu, I'm happy to look over
code with you.  If I understand you correctly, and you are operating in the
mid-tier, it's really easy and much preferred to use a filter to handle
your connection closing.  Putting connections in the session can subject
them to timeouts, etc.

iRODS is designed for short lived agents, actually (think iCommands) and
while there are occasions where long running connections are indicated,
this isn't a good place for that!  As described, we'll soon have an easy
way for you to introduce a pool, and that will include connection health
checks and reasonable time-out management in the background.  I can whip up
some example filters for you in pretty short order, but generally for web
stuff that is a good pattern.

CHeers
MC

**Comment by ngmenyen** [comment]
Hi Mike! My apologies for going silent these days. I actually managed to fix my problem with connection leaks thanks to your advice.

All I did, was add the closeSessionAndEatExceptions() method that you mentioned in one of our handler interceptors. Now all iRODS Agent process seem to be under control. 

I'm still testing the fix, but I don't think this problem will come back any time soon. If you want to talk about the fix, I left my email visible in my profile.

Best regards,
Ng Men Yen
